group/assignment5/find_alphabet.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python3
+
+from collections import deque
+from itertools import combinations
+
+
+def _common_prefix(s1, s2):
+    """ given two strings, returns position of first difference
+            or -1 if strings are the same """
+    for i in range(min(len(s1), len(s2))):
+        if s1[i] != s2[i]:
+            return i
+    return -1
+
+
+def _make_graph(dictionary):
+    """ creates graph of lex order from dictionary """
+    alphabet = set()
+    for word in dictionary:
+        alphabet |= set(word)
+
+    graph = {key: set() for key in alphabet}
+
+    for word1, word2 in zip(dictionary, dictionary[1:]):
+        cp = _common_prefix(word1, word2)
+        if cp != -1:
+            graph[word1[cp]].add(word2[cp])
+
+    return graph
+
+
+def _is_cycle(v, graph, color):
+    """ dfs checking whether cycle exists """
+    color[v] = 1
+    for u in graph[v]:
+        if color[u] == 1:
+            return True
+        elif not color[u] and _is_cycle(u, graph, color):
+            return True
+    color[v] = 2
+    return False
+
+
+def _topsort(v, graph, used, order):
+    """ computes topsort of acyclic graph """
+    used.add(v)
```

Comment:
I think we should try to find a way to find maximal and minimal constraints during topsort -> how can we detect the cycle here?